File: Clocktower-UI/src/types/townUser.ts
import type {
    ClocktowerServerDataTownUser
} from "@/api";
import {
    mapToVoiceState,
    type VoiceState
} from "@/types/voiceState.ts";

export type TownUser = {
    id: string;
    name: string;
    avatarUrl: string | null;
    isPresent: boolean,
    voiceState: VoiceState | null
};

export function mapToTownUser(apiGameUser: ClocktowerServerDataTownUser): TownUser {
    return {
        id: apiGameUser.id!,
        name: apiGameUser.name!,
        avatarUrl: apiGameUser.avatarUrl ?? null,
        isPresent: apiGameUser.isPresent ?? false,
        voiceState: apiGameUser.voiceState ? mapToVoiceState(apiGameUser.voiceState) : null
    };
}

File: Clocktower.Server/Data/TownUser.cs
namespace Clocktower.Server.Data;

public record TownUser(string Id, string Name, string AvatarUrl) : ITownUser
{
    public VoiceState VoiceState { get; set; } = null!;
    public bool IsPresent { get; set; }
}

public interface ITownUser : IIdentifiable
{
    [UsedImplicitly] string Name { get; }
    [UsedImplicitly] string AvatarUrl { get; }
    [UsedImplicitly] VoiceState VoiceState { get; set; }
    [UsedImplicitly] bool IsPresent { get; set; }
}